### PR TITLE
solution

### PR DESCRIPTION
**Pull request by maksym-kostetskyi**
https://maksym-kostetskyi.github.io/react_dynamic-list-of-todos

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,14 +1,48 @@
...
+    setLoading(true);
+
+    getTodos()
+      .then(todosFromServer => {
+        return todosFromServer.filter(todo =>
+          todo.title.toLowerCase().includes(search.toLowerCase()),
+        );
+      })
+      .then(todosBySearch => {
+        if (filter === 'active') {
```
Comment (volodymyr-soltys97) [inline src/App.tsx:34]: Сreate a enum for 'active', 'completed' and 'all' and use it everywhere

File: src/App.tsx [hunk 2]
```diff
@@ -17,18 +51,34 @@ export const App: React.FC = () => {
...
-              <TodoFilter />
+              <TodoFilter filter={setFilter} search={setSearch} />
             </div>
 
             <div className="block">
-              <Loader />
-              <TodoList />
+              {loading && <Loader />}
+
+              {!loading && todos.length > 0 && (
```
Comment (volodymyr-soltys97) [inline src/App.tsx:60]:
```suggestion
              {!loading && !!todos.length && (
```

File: src/App.tsx [hunk 3]
```diff
@@ -17,18 +51,34 @@ export const App: React.FC = () => {
...
+                <TodoList
+                  todos={todos as Todo[]}
+                  setCurrentModal={setCurrentModal}
+                  currentModal={currentModal}
+                />
+              )}
+
+              {!loading && todos.length === 0 && (
```
Comment (volodymyr-soltys97) [inline src/App.tsx:68]:
```suggestion
              {!loading && !todos.length && (
```

File: src/components/TodoList/TodoList.tsx [hunk 4]
```diff
@@ -16,85 +28,46 @@ export const TodoList: React.FC = () => (
...
-            <span className="icon">
-              <i className="far fa-eye" />
-            </span>
-          </button>
-        </td>
-      </tr>
+          <td className="is-vcentered is-expanded">
+            <p
+              className={
+                !todo.completed ? 'has-text-danger' : 'has-text-success'
```
Comment (volodymyr-soltys97) [inline src/components/TodoList/TodoList.tsx:45]: Use the classnames library for add classes with condition

**Review by volodymyr-soltys97 — CHANGES_REQUESTED** [review]
Good job 👍 
Let's make your code better

File: src/components/TodoList/TodoList.tsx [hunk 5]
```diff
@@ -1,100 +1,76 @@
...
-          <button data-cy="selectButton" className="button" type="button">
-            <span className="icon">
-              <i className="far fa-eye" />
-            </span>
-          </button>
-        </td>
-      </tr>
+      <tbody>
+        {todos.map(todo => (
+          <tr data-cy="todo" className="">
```
Comment (anastasiiavorobiova) [inline src/components/TodoList/TodoList.tsx:34]: Consider using keys while rendering lists

**Review by anastasiiavorobiova — CHANGES_REQUESTED** [review]
Great work! Almost done!

File: src/App.tsx [hunk 6]
```diff
@@ -1,14 +1,49 @@
...
 import { Loader } from './components/Loader';
+import { getTodos } from './api';
+import { Todo } from './types/Todo';
+import { Filter } from './types/Filter';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [loading, setLoading] = useState(false);
+  const [filter, setFilter] = useState(Filter.all);
+  const [search, setSearch] = useState('');
```
Comment (GUSILLUS) [inline src/App.tsx:17]: consider renaming to searchQuery to improve readability

File: src/App.tsx [hunk 7]
```diff
@@ -1,14 +1,49 @@
...
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { getTodos } from './api';
+import { Todo } from './types/Todo';
+import { Filter } from './types/Filter';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [loading, setLoading] = useState(false);
```
Comment (GUSILLUS) [inline src/App.tsx:15]: Consider renaming it to isLoading https://dev.to/michi/tips-on-naming-boolean-variables-cleaner-code-35ig

File: src/components/TodoFilter/TodoFilter.tsx [hunk 8]
```diff
@@ -1,30 +1,59 @@
...
-        type="text"
-        className="input"
-        placeholder="Search..."
-      />
-      <span className="icon is-left">
-        <i className="fas fa-magnifying-glass" />
-      </span>
+type Props = {
+  filter: React.Dispatch<Filter>;
+  search: React.Dispatch<string>;
```
Comment (GUSILLUS) [inline src/components/TodoFilter/TodoFilter.tsx:6]: Consider renaming it to clarify that this is some sort of function `changeFilter` `changeSearchQuery`

File: src/components/TodoFilter/TodoFilter.tsx [hunk 9]
```diff
@@ -1,30 +1,59 @@
...
 
-      <span className="icon is-right" style={{ pointerEvents: 'all' }}>
-        {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
-        <button data-cy="clearSearchButton" type="button" className="delete" />
-      </span>
-    </p>
-  </form>
-);
+export const TodoFilter: React.FC<Props> = ({ filter, search }) => {
+  const [input, setInput] = useState('');
```
Comment (GUSILLUS) [inline src/components/TodoFilter/TodoFilter.tsx:10]: You don't need this state, you can use `searchQuery` state here just pass it in props

File: src/components/TodoFilter/TodoFilter.tsx [hunk 10]
```diff
@@ -1,30 +1,59 @@
...
+    <form className="field has-addons">
+      <p className="control">
+        <span className="select">
+          <select
+            data-cy="statusSelect"
+            onChange={e => filter(e.target.value as Filter)}
+          >
+            <option value={Filter.all}>All</option>
+            <option value={Filter.active}>Active</option>
+            <option value={Filter.completed}>Completed</option>
```
Comment (GUSILLUS) [inline src/components/TodoFilter/TodoFilter.tsx:27]: You can use map rendering here with enum values `Object.values(Filter)`

File: src/components/TodoList/TodoList.tsx [hunk 11]
```diff
@@ -1,100 +1,76 @@
...
-          <button data-cy="selectButton" className="button" type="button">
-            <span className="icon">
-              <i className="far fa-eye" />
-            </span>
-          </button>
-        </td>
-      </tr>
+      <tbody>
+        {todos.map(todo => (
+          <tr data-cy="todo" key={todo.id}>
```
Comment (GUSILLUS) [inline src/components/TodoList/TodoList.tsx:34]: Consider creating todo item component and use destruction assignment for todo object

**Review by GUSILLUS — CHANGES_REQUESTED** [review]
GJ

File: src/types/TodoType.ts [hunk 12]
```diff
@@ -1,4 +1,4 @@
-export interface Todo {
+export interface TodoType {
```
Comment (SanyaBratashchuk) [inline src/types/TodoType.ts:1]:
```suggestion
export interface Todo {
```

File: src/components/TodoList/TodoList.tsx [hunk 13]
```diff
@@ -1,100 +1,43 @@
...
-          </button>
-        </td>
-      </tr>
-    </tbody>
-  </table>
-);
+      <tbody>
+        {todos.map(todo => (
+          <Todo
+            todo={todo as TodoType}
```
Comment (SanyaBratashchuk) [inline src/components/TodoList/TodoList.tsx:35]:
```suggestion
            todo={todo}
```

**Review by SanyaBratashchuk — CHANGES_REQUESTED** [review]

**Review by anastasiiavorobiova — APPROVED** [review]
Great work!

**Pull request closed without merging**